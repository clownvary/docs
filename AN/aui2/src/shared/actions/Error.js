import { isSystemError } from '../api/parseResponse';

export const ADD_ERROR = 'ADD_ERROR';
export const CLEAR_ERROR = 'CLEAR_ERROR';

export function addError({ payload }) {
  const code = payload.code;
  return {
    type: ADD_ERROR,
    payload: {
      ...payload,
      isSystemError: isSystemError(code)
    }
  };
}

export function clearError() {
  return {
    type: CLEAR_ERROR
  };
}
