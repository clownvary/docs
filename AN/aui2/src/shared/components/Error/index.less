@error-bar-width: 650px;
/* stylelint-disable */

.error-bar {
  &-body {
    display: block;
    padding: 30px;
    line-height: 1.5em;
    background-color: #fff;
    border-radius: 0 0 10px 10px;

    .error {
      display: flex;
      padding: 0;
      background: #fff;
      border-color: #fff;

      &-content {
        padding-right: 30px;
        padding-left: 10px;
        vertical-align: top;

        ul {
          padding: 0;
          margin: 0;
          list-style-type: none;
        }
      }

      &-reload-btn {
        cursor: pointer;
        color: #009dd9;
      }

      .go-login-btn {
        position: absolute;
        top: 90px;
        right: 0;
      }
    }
  }

  &-button {
    float: right;
    margin-top: 15px;
  }

  .modal-box {
    width: @error-bar-width;
  }
}
