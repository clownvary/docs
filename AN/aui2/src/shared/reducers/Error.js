import { fromJS } from 'immutable';
import reducerHandler from 'shared/utils/reducerHandler';

import {
  ADD_ERROR,
  CLEAR_ERROR
} from '../actions/Error';

const initialState = fromJS({
  list: [],
  systemErrors: [],
  businessErrors: []
});

const handlers = {
  [ADD_ERROR](state, { payload }) {
    let list = state.get('list');
    const exist = list.some(item => item.code === '0002' || (item.code === payload.code && item.message === payload.message));

    if (exist) {
      return state;
    }
    list = list.push({ ...payload });
    const systemErrors = list.filter(err => err.isSystemError);
    const businessErrors = list.filter(err => !err.isSystemError);

    return state.withMutations((s) => {
      s.set('list', list);
      s.set('systemErrors', systemErrors);
      s.set('businessErrors', businessErrors);
    });
  },

  [CLEAR_ERROR](state) {
    return state.withMutations((s) => {
      s.set('list', fromJS([]));
      s.set('systemErrors', fromJS([]));
      s.set('businessErrors', fromJS([]));
    });
  }
};

export default reducerHandler(initialState, handlers);
