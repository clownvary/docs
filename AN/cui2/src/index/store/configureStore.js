import { createStore, applyMiddleware, compose } from 'redux';
import { routerMiddleware } from 'react-router-redux';
import { presetRedux } from 'react-base-ui/lib/middlewares';
import { redirectMiddleware, legacyCuiMiddleware } from 'shared/middlewares';
import rootReducer from 'index/reducers';

export default (initialState = {}, history) => {
  const middlewareList = [
    redirectMiddleware, ...presetRedux, legacyCuiMiddleware, routerMiddleware(history)
  ];

  const composeEnhancers =
    typeof window === 'object' &&
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ ?
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({
      // Specify extensionâ€™s options like name, actionsBlacklist, actionsCreators, serialize...
    }) : compose;

  const middleware = __DEV__ || __STATIC__ ?
   composeEnhancers(applyMiddleware(...middlewareList)) : applyMiddleware(...middlewareList);

  const store = middleware(createStore)(rootReducer, initialState);

  if ((__DEV__ || __STATIC__) && module.hot) {
    module.hot.accept('../reducers', () => {
      store.replaceReducer(require('../reducers')); // eslint-disable-line global-require
    });
  }

  return store;
};
