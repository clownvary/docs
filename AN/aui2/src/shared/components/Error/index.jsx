import React from 'react';
import { connect } from 'react-redux';
import Button from 'react-base-ui/lib/components/Button';
import Modal from 'react-base-ui/lib/components/Modal';
import * as da from 'react-base-ui/lib/utils/dataAccess';
import { clearError } from 'shared/actions/Error';
import { redirect, reload as locationReload } from 'shared/actions/route';
import UIComponent from 'shared/components/UIComponent';
import { isInIframe } from 'shared/utils/iframe';

import './index.less';

export class Error extends UIComponent {
  constructor(props, context) {
    super(props, context);
    this.bind('onClick', 'gotoLogin', 'reload');
  }

  displayError() {
    const { error } = this.props;
    const errorList = da.get(error, 'list');
    return !!errorList.size;
  }

  sessionTimeoutError() {
    const { error } = this.props;
    const errorList = da.get(error, 'list');
    return !!errorList.find(item => item && item.code === '0002');
  }

  render() {
    const { title = 'System Message', reload = false } = this.props;
    const display = this.displayError();

    return (
      <Modal shown={display} title={title} className="error-bar" onClose={this.onClick}>
        <div className="error-bar-body u-clearfix">
          {this.renderErrorAlert()}
          <Button size="sm" type="strong" className="error-bar-button" onClick={this.onClick}>
            {reload ? 'Reload Page' : 'OK'}
          </Button>
        </div>
      </Modal>
    );
  }

  renderErrorAlert() {
    const { error } = this.props;
    const errorList = da.get(error, 'list');
    return (
      <div className="error-content">
        <ul>
          {errorList.map((item, k) => <li key={k}>{item.message}</li>)}
        </ul>
      </div>
    );
  }

  reload() {
    /* istanbul ignore next */
    if (isInIframe) {
      this.props.locationReload(window.frameElement.contentWindow);
    } else {
      this.props.locationReload();
    }
  }

  gotoLogin() {
    /* istanbul ignore next */
    if (!__STATIC__) {
      const homeUrl = `${window.__environment__.ROOT_URL}/servlet/adminlogin.sdi?sessiontimeout=Checked`;
      this.props.redirect(homeUrl);
    } else {
      this.reload();
    }
  }

  onClick() {
    const { reload } = this.props;
    this.props.clearError();
    if (this.sessionTimeoutError()) {
      this.gotoLogin();
    } else if (reload) {
      this.reload();
    }
  }
}

export default connect(
  null,
  {
    clearError,
    redirect,
    locationReload
  }
)(Error);
