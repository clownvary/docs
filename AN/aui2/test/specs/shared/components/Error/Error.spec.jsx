import React from 'react';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import { mount } from 'enzyme';
import Error from 'shared/components/Error';

describe('shared/components/Error', () => {

  let store;

  beforeEach(() => {
    const mockStore = configureStore();
    store = mockStore();
  });

  afterEach(() => {
    store.clearActions();
  });

  const setup = props => mount(
    <Error display {...props} store={store} />
  );

  it('Error component should render correctly if no error', () => {
    const props = {
      error: {
        list: []
      }
    };
    const component = setup(props);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('Error component should render correctly with errors', () => {
    const props = {
      title: 'Error for snapshot',
      error: {
        list: [
          {
            message: 'error message1'
          },
          {
            message: 'error message2'
          },
          {
            message: 'error message3'
          },
        ]
      }
    };

    const component = setup(props);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('Error component should work fine if no reload require with normal error', () => {
    const props = {
      error: {
        list: [
          {
            message: 'error message'
          }
        ]
      }
    };

    const component = setup(props);

    component.find('button').simulate('click');
  });

  it('Error component should work fine if reload required with normal error', () => {
    const props = {
      reload: true,
      error: {
        list: [
          {
            message: 'error message'
          }
        ]
      }
    };

    const component = setup(props);

    component.find('button').simulate('click');
  });

  it('Error component should work fine with session timeout error', () => {
    const props = {
      error: {
        list: [
          {
            code: '0002',
            message: 'error message'
          }
        ]
      }
    };

    const component = setup(props);

    component.find('button').simulate('click');
  });
});
