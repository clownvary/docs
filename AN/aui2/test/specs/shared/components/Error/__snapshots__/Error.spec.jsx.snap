// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shared/components/Error Error component should render correctly if no error 1`] = `
<Connect(Error)
  display={true}
  error={
    Object {
      "list": Array [],
    }
  }
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "subscribe": [Function],
    }
  }
>
  <Error
    clearError={[Function]}
    display={true}
    error={
      Object {
        "list": Array [],
      }
    }
    locationReload={[Function]}
    redirect={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "subscribe": [Function],
      }
    }
  >
    <Modal
      className="error-bar"
      onClose={[Function]}
      shown={false}
      title="System Message"
    >
      <div
        className="modal-wrap"
        role="dialog"
      >
        <div
          className="modal error-bar"
          onKeyDown={[Function]}
        >
          <section
            className="modal-box"
          >
            <header
              className="modal-header"
            >
              <h3
                className="modal-title"
              >
                System Message
              </h3>
              <span
                aria-label="close-dialog"
                className="icon icon-close modal-close"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="button"
                tabIndex="0"
              />
            </header>
            <div
              className="error-bar-body u-clearfix"
            >
              <div
                className="error-content"
              >
                <ul />
              </div>
              <Button
                className="error-bar-button"
                loading={false}
                menuData={Array []}
                noSubmit={false}
                onClick={[Function]}
                size="sm"
                type="strong"
              >
                <button
                  className="btn btn-strong btn--sm error-bar-button"
                  onClick={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  onMouseOver={[Function]}
                  type="submit"
                >
                  <span>
                    OK
                  </span>
                </button>
              </Button>
            </div>
          </section>
        </div>
        <div
          className="modal-mask"
        />
      </div>
    </Modal>
  </Error>
</Connect(Error)>
`;

exports[`shared/components/Error Error component should render correctly with errors 1`] = `
<Connect(Error)
  display={true}
  error={
    Object {
      "list": Array [
        Object {
          "message": "error message1",
        },
        Object {
          "message": "error message2",
        },
        Object {
          "message": "error message3",
        },
      ],
    }
  }
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "subscribe": [Function],
    }
  }
  title="Error for snapshot"
>
  <Error
    clearError={[Function]}
    display={true}
    error={
      Object {
        "list": Array [
          Object {
            "message": "error message1",
          },
          Object {
            "message": "error message2",
          },
          Object {
            "message": "error message3",
          },
        ],
      }
    }
    locationReload={[Function]}
    redirect={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "subscribe": [Function],
      }
    }
    title="Error for snapshot"
  >
    <Modal
      className="error-bar"
      onClose={[Function]}
      shown={false}
      title="Error for snapshot"
    >
      <div
        className="modal-wrap"
        role="dialog"
      >
        <div
          className="modal error-bar"
          onKeyDown={[Function]}
        >
          <section
            className="modal-box"
          >
            <header
              className="modal-header"
            >
              <h3
                className="modal-title"
              >
                Error for snapshot
              </h3>
              <span
                aria-label="close-dialog"
                className="icon icon-close modal-close"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="button"
                tabIndex="0"
              />
            </header>
            <div
              className="error-bar-body u-clearfix"
            >
              <div
                className="error-content"
              >
                <ul>
                  <li
                    key="0"
                  >
                    error message1
                  </li>
                  <li
                    key="1"
                  >
                    error message2
                  </li>
                  <li
                    key="2"
                  >
                    error message3
                  </li>
                </ul>
              </div>
              <Button
                className="error-bar-button"
                loading={false}
                menuData={Array []}
                noSubmit={false}
                onClick={[Function]}
                size="sm"
                type="strong"
              >
                <button
                  className="btn btn-strong btn--sm error-bar-button"
                  onClick={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  onMouseOver={[Function]}
                  type="submit"
                >
                  <span>
                    OK
                  </span>
                </button>
              </Button>
            </div>
          </section>
        </div>
        <div
          className="modal-mask"
        />
      </div>
    </Modal>
  </Error>
</Connect(Error)>
`;
